### PR TITLE
GH-2744: ScatterGather: reinstate request headers

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration/issues/2744

When we get scattering results, there is no reason to keep internal
headers any more.
* Fix `ScatterGatherHandler` to modify scattering result messages to
reinstate headers from original request message.
This way we are able to re-throw an exception from the gatherer to
the caller

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/scattergather/ScatterGatherHandler.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/scattergather/ScatterGatherHandler.java]: `s/delivery/deliver the/`

File: src/reference/asciidoc/scatter-gather.adoc
Comment (garyrussell) [inline src/reference/asciidoc/scatter-gather.adoc]: `s/if async/if an async/`

`s/applied for scatter recipients/applied in scatter recipient subflows/` ?

File: src/reference/asciidoc/scatter-gather.adoc
Comment (garyrussell) [inline src/reference/asciidoc/scatter-gather.adoc]: waiting for **a** reply ... forever,  by default

**Review by garyrussell — CHANGES_REQUESTED** [review]
A few minor nit-picks

**Comment by artembilan** [comment]
Gary, requested changes have been pushed.
There is a `ChannelInterceptorAware` point in the `ScatterGatherHandler`, which has to be fixed after merging #2751 .
The fix plan depends of the order of merging 😄 

Thanks

File: src/reference/asciidoc/scatter-gather.adoc
Comment (garyrussell) [inline src/reference/asciidoc/scatter-gather.adoc]: an async - will fix on merge

**Comment by garyrussell** [comment]
Merged as 0d37566452725d04bc92793c134be480e2905843

**Pull request closed without merging**
